python-libraries/data-platform-catalogue/data_platform_catalogue/client/datahub/datahub_client.py

```diff
@@ -151,41 +151,51 @@ def upsert_data_product(self, metadata: DataProductMetadata):
 
         data_product_urn = "urn:li:dataProduct:" + "".join(name.split())
 
-        if metadata.domain:
-            domain = metadata_dict.pop("domain")
-            domain_urn = self.graph.get_domain_urn_by_name(domain_name=domain)
+        domain = metadata_dict.pop("domain")
+        domain_urn = self.graph.get_domain_urn_by_name(domain_name=domain)
 
-            if domain_urn is None:
-                logger.info(f"creating new domain {domain} for {name}")
-                domain_urn = self.create_domain(domain=domain)
+        if domain_urn is None:
+            logger.info(f"creating new domain {domain} for {name}")
+            domain_urn = self.create_domain(domain=domain)
 
-            data_product_domain = DomainsClass(domains=[domain_urn])
-            metadata_event = MetadataChangeProposalWrapper(
-                entityType="dataproduct",
-                changeType=ChangeTypeClass.UPSERT,
-                entityUrn=data_product_urn,
-                aspect=data_product_domain,
-            )
-            self.graph.emit(metadata_event)
-            logger.info(f"Data Product {name} associated with domain {domain}")
+        subdomain_urn = None
+        if metadata.subdomain:
```

Comment:
All of the `domain` logic should be an `elif`/`else` for if this conditional is false, as you're not using `domain`/`domain_urn` if `subdomain` exists.

Reply:
Nope this is wrong. Please ignore. I missed that you're creating a parent-child relationship for any new subdomain, which requires the `domain_urn`

Reply:
domain & domain_urn is needed when subdomain is present in the metadata but not yet created, so there would still be duplication of the calls. Happy to chat about moving to another ticket.

Reply:
@tom-webber I think if subdomain is passed in we may still create the domain if it doesn't exist - otherwise we can't use it as the parent of the subdomain...

I think the scenarios are something like this:

| | domain passed in without subdomain | domain & subdomain passed in | 
| -- | -- | -- | 
| subdomain exists | n/a | do nothing |
| domain exists | do nothing  | create subdomain |
| nothing exists | create domain | create domain and subdomain |